rfcs/rfc-HAARG-0001.md

```diff
@@ -0,0 +1,180 @@
+maps are needed. You need to manually save the `$_` value in another variable.
+
+```perl
+my %hash = (
+  foo => [ 1, 2, 3, 4 ],
+  bar => [ 5, 6, 7, 8 ],
+);
+
+my @out = map {
+  my $key = $_;
+  map { "$key: $_" } $hash{$key}->@*;
+} keys %hash;
+```
+
+Using `$_` can also be dangerous if you need to call code not under your
+direct control, due to it being the implicit target of operations like
+`readline`.
+
+It would be more convenient if you could use a different variable for the
+topic, similar to what `for` allows.
+
+```perl
+my @out = map my $key {
+  map my $i { "$key: $i" } $hash{$key}->@*;
+} keys %hash;
+```
+
+A natural extension of this syntax would be to allow n-at-a-time iteration, as
+`for` can do on perl 5.36+.
+
+```perl
+my @out = map my ($key, $val) {
+  map my $i { "$key: $i" } $val->@*;
+} %hash;
+```
+
+All of this would apply to `grep` as well.
+
+As the syntax proposed is currently invalid, a feature should not be needed to
+enable it.
+
+## Rationale
+
+The syntax chosen is meant to follow from the syntax of `for`, treating `map`
+as "`for` as an expression". The chosen syntax does not create any
+ambiguities, and naturally extends to follow the syntax used by `for` for
+n-at-a-time iteration.
+
+## Specification
+
+### `map my VAR BLOCK LIST`
+
+This will evaluate `BLOCK` for each element of `LIST`, aliasing `VAR` to each
+element. Its behavior will otherwise match `map BLOCK LIST`. It is only
+possible to use `my` variables for this.
+
+### `map my (VAR, VAR) BLOCK LIST`
+
+This will evaluate `BLOCK` for each set of two elements in `LIST`, aliasing
+the `VAR` to the first of each set, and the second `VAR` to the second. If
+there is not a multiple of two items in `LIST`, the last run of `BLOCK` will
+assign `undef` to the extra `VAR` slot not corresponding to an element in
+`LIST`. More than two variables can be used to iterate over sets of three or
+more items.
+
+## Backwards Compatibility
+
+As the syntax chosen is currently invalid, it should not present any backwards
+compatibility concerns.
+
+While the parsing of `map` and `grep` is complex and has ambiguities, the
+additions proposed are bounded and simple to parse, so they do not introduce
+any new ambiguities.
+
+## Security Implications
+
+None foreseen.
+
+## Examples
+
+```perl
+my @my_array = (11 .. 14);
+
+my @trad_map = map { $_ + 1 } @my_array;
+# ( 12, 13, 14, 15 )
+ 
+my @new_map = map my $f { $f + 1 } @my_array;
+# ( 12, 13, 14, 15 )
+
+my @pair_map = map my ($f, $g) { $f + $g } @my_array;
+# ( 23, 27 )
+
+my @new_grep = grep my $f { $f > 12 } @my_array;
+# ( 13, 14 )
+
+my %my_hash = (
+   first_key => 11,
+   second_key => 14,
+);
+
+my %pair_grep = grep my ($f, $g) { $f =~ /sec/ } %my_hash;
+# ( second_key => 14 )
+```
+
+## Prototype Implementation
+
+  - List::Util includes the functions `pairmap` and `pairgrep` which can do
+    n-at-a-time iteration. These use the "magic" `$a` and `$b` variables.
```

Comment:
Only assuming n=2.

Reply:
Yes; the pair* functions are "n-at-a-time" for specifically n=2; the ones suggested by this RFC are even more general than that.